core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java

```diff
@@ -0,0 +1,298 @@
+   * buffer's state.
+   *
+   * @param journalEntry the journal entry
+   */
+  public void processJournalEntry(Journal.JournalEntry journalEntry) {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      if (mSuspended) {
+        // New entry submissions are monitored by catch-up threads.
+        synchronized (mSuspendBuffer) {
+          mSuspendBuffer.offer(journalEntry);
+          mSuspendBuffer.notifyAll();
+        }
+      } else {
+        applyToMaster(journalEntry);
+      }
+    }
+  }
+
+  /**
+   * @return {@code true} if this applier was suspended
+   */
+  public boolean isSuspended() {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      return mSuspended;
+    }
+  }
+
+  /**
+   * Suspend the applier.
+   *
+   * After this call, journal entries will be buffered until {@link #resume()} or
+   * {@link #catchup(long)} is called.
+   *
+   * @throws IOException
+   */
+  public void suspend() throws IOException {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      Preconditions.checkState(!mSuspended, "Already suspended");
+      LOG.info("Suspending state machine.");
+      mSuspended = true;
+      LOG.info("Suspended state machine at sequence: {}", mLastAppliedSequence);
+    }
+  }
+
+  /**
+   * Resumes the applier. This method will apply all buffered entries before returning.
+   *
+   * @throws IOException
+   */
+  public void resume() throws IOException {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      Preconditions.checkState(mSuspended, "Not suspended");
+      Preconditions.checkState(!mResumeInProgress, "Resume in progress");
+      mResumeInProgress = true;
+      LOG.info("Resuming state machine.");
+    }
+
+    // Cancel catching up thread if active.
+    if (mCatchupThread != null && mCatchupThread.isAlive()) {
+      mCatchupThread.cancel();
+      mCatchupThread.waitTermination();
+    }
+
+    /**
+     * Applies all buffered entries.
+     *
+     * It doesn't block state until
+     *   -> buffer contains few elements ( RESUME_LOCK_BUFFER_SIZE_WATERMARK )
+     *   -> was running for a long time  ( RESUME_LOCK_TIME_LIMIT_MS         )
+     */
+    try {
+      // Mark resume start time.
+      Instant resumeStartTime = Instant.now();
+      // Lock initially if few or none elements in the queue.
+      if (mSuspendBuffer.size() <= RESUME_LOCK_BUFFER_SIZE_WATERMARK) {
+        mStateLock.lock();
+      }
+
+      while (mSuspendBuffer.size() > 0) {
+        applyToMaster(mSuspendBuffer.remove());
+
+        // Check whether to lock the state now.
+        boolean lockSubmission = !mStateLock.isHeldByCurrentThread()
+            && (mSuspendBuffer.size() <= RESUME_LOCK_BUFFER_SIZE_WATERMARK
+                || Duration.between(Instant.now(), resumeStartTime)
+                    .compareTo(Duration.ofMillis(RESUME_LOCK_TIME_LIMIT_MS)) > 0);
```

Comment:
Is there a downside to always grabbing the lock and flushing the entire queue? In the worst case scenario, this is what we revert to after waiting `RESUME_LOCK_TIME_LIMIT`

Reply:
This is to not prevent `processJournalEntry` if buffer was overloaded during a long backup. Blocking it for a long time would render copycat server unreachable.

Reply:
Could we still be in that situation after `RESUME_LOCK_TIME_LIMIT` is up?

Reply:
Right. Hitting a time limit is when the code stops being hopeful. Hitting a time limit is extremely unlikely because resuming is done on a thread, where as new entries will come in as part of copycat hearbeats.